In [3]:
import numpy as np
def load_data(file="small.txt"):
    fh = open(file, "r")
    dataset = [[*x.strip()] for x in fh.readlines()]
    dataset = np.array(dataset)
    return dataset

In [12]:
dataset = load_data()

In [68]:

def north(dataset):
    for col in range(dataset.shape[1]):
        d_row = 0
        for row in range(dataset.shape[0]):
            c = dataset[row, col]
            if c == ".":
                # "." move up atleast one:
                d_row += 1
            elif c == "#":
                # Nothing to move:
                d_row = 0
            elif c == "O":
                # Move "O" by delta:
                dataset[row, col] = "."
                dataset[row - d_row, col] = "O"
    return dataset

def south(dataset):
    for col in range(dataset.shape[1]):
        d_row = 0
        for row in reversed(range(dataset.shape[0])):
            c = dataset[row, col]
            if c == ".":
                # "." move up atleast one:
                d_row += 1
            elif c == "#":
                # Nothing to move:
                d_row = 0
            elif c == "O":
                # Move "O"
                dataset[row, col] = "."
                dataset[row + d_row, col] = "O"
    return dataset

def east(dataset):
    for row in range(dataset.shape[0]):
        d_col = 0
        for col in reversed(range(dataset.shape[1])):
            c = dataset[row, col]
            if c == ".":
                # "." move up atleast one:
                d_col += 1
            elif c == "#":
                # Nothing to move:
                d_col = 0
            elif c == "O":
                # Move "O"
                dataset[row, col] = "."
                dataset[row, col+d_col] = "O"
    return dataset

def west(dataset):
    for row in range(dataset.shape[0]):
        d_col = 0
        for col in range(dataset.shape[1]):
            c = dataset[row, col]
            if c == ".":
                # "." move up atleast one:
                d_col += 1
            elif c == "#":
                # Nothing to move:
                d_col = 0
            elif c == "O":
                # Move "O"
                dataset[row, col] = "."
                dataset[row, col-d_col] = "O"
    return dataset


In [69]:
dataset = load_data()
dataset = west(dataset)

for row in dataset:
    print("".join(row))

O....#....
OOO.#....#
.....##...
OO.#OO....
OO......#.
O.#O...#.#
O....#OO..
O.........
#....###..
#OO..#....


In [70]:
import numpy as np
def load_data(file="small.txt"):
    fh = open(file, "r")
    dataset = [[*x.strip()] for x in fh.readlines()]
    dataset = np.array(dataset)
    return dataset

def rotate(dataset):
    dataset = north(dataset)
    dataset = west(dataset)
    dataset = south(dataset)
    dataset = east(dataset)
    return dataset

def score(dataset):
    score = 0
    for r, row in enumerate(dataset):
        for c, _ in enumerate(row):
            if dataset[r, c] == "O":
                score += dataset.shape[0] - r
    return score

In [74]:
dataset = load_data("big.txt")
cache = dict()
todo = 0
TOTAL_CYCLES = 1000000000

for cycle in range(TOTAL_CYCLES):
    dataset = rotate(dataset)

    # Need a key. Use tuple in this case:
    key = tuple(tuple(x) for x in dataset)
    if key in cache:
        diff = cycle - cache[key]
        todo = (TOTAL_CYCLES - cycle) % diff - 1
        print(f"Only have to do {cycle} cycles.")
        break
    cache[key] = cycle

for _ in range(todo):
    dataset = rotate(dataset)
score(dataset)

Only have to do 159 cycles.


105606

- Only have to do 9 cycles. Running final set of 3 cycles...
- Solution:  64

- Only have to do 159 cycles. Running final set of 60 cycles...
- Solution:  105606

In [32]:
for key in cache:
    if cache[key] == (9 - 3 - 1):
        print(cache[key], score(np.array(key)))

5 64
